File: cpp/se/LoginCommand.cpp
#include "se/LoginCommand.hpp"
#include "se/StandardCommandType.hpp"
#include "se/StandardObjectType.hpp"
#include "xml/XMLHelper.hpp"

#include <xercesc/dom/DOMElement.hpp>

using namespace std;

const string LoginCommand::DEFAULT_VERSION("1.0");
const string LoginCommand::DEFAULT_LANG("en");

LoginCommand::LoginCommand(const string &clID,
                           const string &password)
    : Command (StandardCommandType::LOGIN())
{
    const vector<string>& stdURIs = StandardObjectType::getStandardURIs();
    
    init(clID, password, DEFAULT_VERSION, DEFAULT_LANG, NULL, stdURIs, vector<string>());
}


LoginCommand::LoginCommand(const string& clID,
                           const string& password,
                           const string* newPW)
    : Command(StandardCommandType::LOGIN())
{
    const vector<string>& stdURIs = StandardObjectType::getStandardURIs();
    
    init(clID, password, DEFAULT_VERSION, DEFAULT_LANG, newPW, stdURIs, vector<string>());
}


LoginCommand::LoginCommand(const string& clID,
                           const string& password,
                           const vector<string> objURIs,
                           const vector<string> extURIs)
    : Command(StandardCommandType::LOGIN())
{
    init(clID, password, DEFAULT_VERSION, DEFAULT_LANG, NULL, objURIs, extURIs);
}                            

LoginCommand::LoginCommand (const string& clID,
                            const string& password,
                            const string& version,
                            const string& lang,
                            const vector<string> objURIs,
                            const vector<string> extURIs)
    : Command(StandardCommandType::LOGIN())
{
    init(clID, password, version, lang, NULL, objURIs, extURIs);
}

LoginCommand::LoginCommand (const string& clID,
                            const string& password,
                            const string* newPassword,
                            const string& version,
                            const string& lang,
                            const vector<string> objURIs,
                            const vector<string> extURIs)
    : Command(StandardCommandType::LOGIN())
{
    init(clID, password, version, lang, newPassword, objURIs, extURIs);
}

void LoginCommand::init(const string& clID, 
						const string& password,
						const string& version, 
						const string& lang,
						const string* newPassword,
						const vector<string> objURIs,
						const vector<string> extURIs)
{
	XMLHelper::setTextContent
		(xmlWriter->appendChild (cmdElement, "clID"), clID);
	XMLHelper::setTextContent
		(xmlWriter->appendChild (cmdElement, "pw"), password);
    
    if (newPassword)
		XMLHelper::setTextContent
			(xmlWriter->appendChild(cmdElement, "newPW"), *newPassword);

    DOMElement *options = xmlWriter->appendChild(cmdElement, "options");
            
	XMLHelper::setTextContent 
		(xmlWriter->appendChild(options, "version"), version);
	XMLHelper::setTextContent
		(xmlWriter->appendChild(options, "lang"), lang);
    
    if (objURIs.size() > 0)
    {
        DOMElement *svcs = xmlWriter->appendChild(cmdElement, "svcs");
        for (unsigned int i = 0; i < objURIs.size(); i++)
		{
			XMLHelper::setTextContent(xmlWriter->appendChild(svcs, "objURI"), objURIs[i]);
		}

        if (extURIs.size() > 0)
        {
            DOMElement *svcExtension = 
                xmlWriter->appendChild (svcs, "svcExtension");
            
            for (unsigned int i = 0; i < extURIs.size(); i++)
				XMLHelper::setTextContent
					(xmlWriter->appendChild (svcExtension, "extURI"),
					 extURIs[i]);
        }
    }
}

File: src/main/java/com/ausregistry/jtoolkit2/se/LoginCommand.java
package com.ausregistry.jtoolkit2.se;

import com.ausregistry.jtoolkit2.ErrorPkg;

import org.w3c.dom.Element;

/**
 * Use this to open an EPP session in order to perform commands only permitted
 * from within the context of a session.  Instances of this class generate, via
 * the toXML method, login service elements compliant with the login
 * specification in RFC5730.
 *
 * @see com.ausregistry.jtoolkit2.se.Greeting For services available to be used
 * in the login command on the chosen EPP server.
 *
 * @see com.ausregistry.jtoolkit2.se.LogoutCommand To end a session opened
 * using LoginCommand.
 */
public final class LoginCommand extends Command {
    private static final long serialVersionUID = -704642499677315944L;

    private static final String DEFAULT_VERSION = "1.0";
    private static final String DEFAULT_LANG = "en";

    public LoginCommand(String clID, String password) {
        this(clID, password, StandardObjectType.getStandardURIs(), null);
    }

    public LoginCommand(String clID, String password, String newPW) {
        this(clID, password, newPW, DEFAULT_VERSION, DEFAULT_LANG,
                StandardObjectType.getStandardURIs(), null);
    }

    public LoginCommand(String clID, String password,
            String[] objURIs, String[] extURIs) {
        this(clID, password, DEFAULT_VERSION, DEFAULT_LANG, objURIs, extURIs);
    }

    public LoginCommand(String clID, String password,
            String version, String lang,
            String[] objURIs, String[] extURIs) {
        this(clID, password, null, version, lang, objURIs, extURIs);
    }

    /**
     *
     * @param clID Required.
     * @param password Required.
     * @param newPassword Optional.
     * @param version Required.
     * @param lang
     * @param objURIs Required.
     * @param extURIs Optional.
     * @throws IllegalArgumentException if {@code clID}, {@code password} or {@code version} are {@code null}.
     */
    public LoginCommand(String clID, String password, String newPassword,
            String version, String lang,
            String[] objURIs, String[] extURIs) {
        super(StandardCommandType.LOGIN);

        if (clID == null || password == null || version == null) {
            throw new IllegalArgumentException(
                    ErrorPkg.getMessage("se.login.missing_arg"));
        }

        xmlWriter.appendChild(cmdElement, "clID").setTextContent(clID);
        xmlWriter.appendChild(cmdElement, "pw").setTextContent(password);
        if (newPassword != null) {
            xmlWriter.appendChild(
                    cmdElement, "newPW").setTextContent(newPassword);
        }
        Element options = xmlWriter.appendChild(cmdElement, "options");
        xmlWriter.appendChild(options, "version").setTextContent(version);
        xmlWriter.appendChild(options, "lang").setTextContent(lang);
        Element svcs = xmlWriter.appendChild(cmdElement, "svcs");
        for (String objURI : objURIs) {
            xmlWriter.appendChild(svcs, "objURI").setTextContent(objURI);
        }

        if (extURIs == null || extURIs.length == 0) {
            return;
        }

        Element svcExtension = xmlWriter.appendChild(svcs, "svcExtension");
        for (String extURI : extURIs) {
            xmlWriter.appendChild(
                    svcExtension, "extURI").setTextContent(extURI);
        }
    }
}
